extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js

```diff
@@ -0,0 +1,187 @@
+    this.assistanceElement_ = dev().assertElement(assistanceElement);
+
+    /** @const @private {JsonObject|null|undefined} */
+    this.configJson_ = tryParseJson(this.assistanceElement_.textContent, e => {
+      throw user().createError(
+          'Failed to parse "amp-viewer-assistance" JSON: ' + e);
+    });
+
+    /** @private @const {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(ampdoc);
+
+    /** @private @const {!../../../src/service/action-impl.ActionService} */
+    this.action_ = Services.actionServiceForDoc(this.assistanceElement_);
+
+    /** @private @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = Services.vsyncFor(ampdoc.win);
+  }
+
+  /**
+   * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   * @return {?Promise}
+   * @private
+   */
+  actionHandler_(invocation) {
+    const {method, args} = invocation;
+    if (method == 'updateActionState' && !!args) {
+      this.viewer_.sendMessageAwaitResponse(method, args).catch(error => {
+        user().error(TAG, error.toString());
+      });
+    } else if (method == 'signIn') {
+      this.requestSignIn_();
+    }
+
+    return null;
+  }
+
+  /**
+   * @private
+   * @restricted
+   */
+  start_() {
+    if (!this.enabled_) {
+      user().info(
+          TAG, 'Invalid AMP Action - no "id=amp-viewer-assistance" element');
+      return this;
+    }
+    return this.viewer_.isTrustedViewer().then(isTrustedViewer => {
+      if (!isTrustedViewer &&
+         !isExperimentOn(this.ampdoc_.win, 'amp-viewer-assistance-untrusted')) {
+        this.enabled_ = false;
+        user().info(TAG, 'Disabling AMP Action since viewer is not trusted');
+        return this;
+      }
+      this.action_.installActionHandler(
+          this.assistanceElement_, this.actionHandler_.bind(this),
+          ActionTrust.HIGH);
+
+      this.getIdTokenPromise();
+
+      this.viewer_.sendMessage('viewerAssistanceConfig',dict({
+        'config': this.configJson_,
+      }));
+      return this;
+    });
+  }
+
+  /**
+   * @return {!Promise<undefined>}
+   */
+  getIdTokenPromise() {
+    return this.viewer_.sendMessageAwaitResponse('getAccessTokenPassive', dict({
+      // For now there's only 1 provider option, so we just hard code it
+      'providers': [GSI_TOKEN_PROVIDER],
+    }))
+        .then(token => {
+          this.setIdTokenStatus_(Boolean(token));
+          return token;
+        }).catch(() => {
+          this.setIdTokenStatus_(/*available=*/false);
+        });
+  }
+
+  /**
+   * @private
+   */
+  requestSignIn_() {
+    this.viewer_.sendMessageAwaitResponse('requestSignIn', dict({
+      'providers': [GSI_TOKEN_PROVIDER],
+    })).then(token => {
+      user().info(TAG, 'Token: ' + token);
+      if (token) {
+        this.setIdTokenStatus_(/*available=*/true);
+        this.action_.trigger(
+            this.assistanceElement_, 'signedIn', null, ActionTrust.HIGH);
+      }
```

Comment:
Do we intentionally not set token status to unavailable on sign-in failure?

Reply:
The token status is set to unavailable by default and then switched on in this request. Since there is not a way to have a token be available => unavailable, there was never a need for the failure case. Regardless, adding in the other case to be thorough.